storage: Rebalance replicas based on disk usage and write load

a-robinson opened the pull request:
This merits more testing, including cleaning up a bunch of the unit tests assuming the approach looks reasonable to you guys, but initial allocsim testing looks good so at least it isn't completely off base.

Fixes https://github.com/cockroachdb/cockroach/issues/15988
Fixes https://github.com/cockroachdb/cockroach/issues/12996

a-robinson committed:
storage: Keep track of per-store replica percentile data

a-robinson committed:
storage: Rebalance replicas based on disk usage and write load

Include a setting to disable the new stats-based rebalancing. All
the old tests still pass, so we can be pretty confident that the
behavior hasn't been changed when the setting is disabled.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16878)
<!-- Reviewable:end -->

a-robinson commented:
Ah, I forgot @BramGruneir is gonna be out of town for a while. Any eyes people want to donate to this in his absence are welcome, but it can reasonably wait. Getting more test cycles in with it would be nice, but it's not blocking anything.

petermattis commented:
Review status: 0 of 18 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/storage/allocator_scorer.go, line 621 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-KoOPp4oq5Mg_rP2bfGk:-KoOPp4oq5Mg_rP2bfGl:b9kuc1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58a7038f9f572ea91bd355fa8ad81e37ef9967/pkg/storage/allocator_scorer.go#L621)):*
> ```Go
> // TODO(a-robinson): Give balanceScore a type alias to discourage direct usage
> // of them except through helper functions?
> func balanceScore(sl StoreList, store roachpb.StoreDescriptor, rangeInfo RangeInfo) float64 {
> ```

I'm having trouble coming up for intuition for how this behaves. Specifically, `balanceContribution` is a bit ad-hoc with a lot of magic values. I don't have anything better to offer at the moment, though.

Perhaps it would help to have some table driven tests that would demonstrate how `balanceScore` changes as one of the dimensions changes. Even better if we could figure out how to visualize this. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
I'll have some time tonight to take a look.


On Jul 6, 2017 4:08 PM, "Peter Mattis" <notifications@github.com> wrote:

> Review status: 0 of 18 files reviewed at latest revision, 1 unresolved
> discussion, some commit checks failed.
> ------------------------------
>
> *pkg/storage/allocator_scorer.go, line 621 at r2
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-KoOPp4oq5Mg_rP2bfGk:-KoOPp4oq5Mg_rP2bfGl:b9kuc1u>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/4e58a7038f9f572ea91bd355fa8ad81e37ef9967/pkg/storage/allocator_scorer.go#L621>):*
>
> // TODO(a-robinson): Give balanceScore a type alias to discourage direct usage// of them except through helper functions?func balanceScore(sl StoreList, store roachpb.StoreDescriptor, rangeInfo RangeInfo) float64 {
>
> I'm having trouble coming up for intuition for how this behaves.
> Specifically, balanceContribution is a bit ad-hoc with a lot of magic
> values. I don't have anything better to offer at the moment, though.
>
> Perhaps it would help to have some table driven tests that would
> demonstrate how balanceScore changes as one of the dimensions changes.
> Even better if we could figure out how to visualize this.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/16878>*
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/16878#issuecomment-313505059>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABihudcT78X9CNoXi1AryWaRrltz42MSks5sLT6ngaJpZM4OPtRs>
> .
>

BramGruneir commented:
So the rebalance score, as @petermattis said, is fairly confusing.  

The scorer was originally designed (well, converted from the rule solver) to have a collection of smaller functions calculating different scores for different aspects, but this seems to drop a whole bunch of them together.  The original design had everything bunched together and it was, with very little data, quite confusing to reason about.  So one of the main goals has to be traceability to understand why a rebalance decision was made.  

Why not use that instead of pushing everything down into a single balance score?  Add the percentages to the candidate struct, and write a collection of small functions that look for underfull, overfull, etc there. 

And are those percentages going to be more helpful than just trying to get the mean.  I think they might just be overly complicating the issue.  But I wouldn't know without some testing.

---

Review status: 0 of 18 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-:-KoYKnGEuxBuT-tDbgs-:b-a4j54g)*
<!-- Sent from Reviewable.io -->

petermattis commented:
@a-robinson Let's discuss in person today.

spencerkimball commented on pkg/roachpb/metadata.proto:
```diff
@@ -92,6 +92,16 @@ message RangeDescriptor {
 }
 
+// Percentiles contains a handful of hard-coded percentiles meant to summarize
+// some sort of distribution.
```
s/some sort of/the/

spencerkimball commented on pkg/storage/allocator.go:
```diff
@@ -43,10 +43,10 @@ const (
-	// the mean range/lease count that permits rebalances/lease-transfers away from
-	// that store.
-	baseRebalanceThreshold = 0.05
+	// baseleaseRebalanceThreshold is the minimum ratio of a store's lease surplus
```
s/baselease/baseLease/

a-robinson commented:
I'm still playing around with some different `percentileScore` numbers and logic, so don't bother reviewing yet, but in the meantime, my notes from chatting with @petermattis yesterday on how to make this a little more palatable:

* create graphs for each dimension for each store
* add a knob to disable non-range-count dimensions
* put debug info into range log
* moar tests
* (future work) incorporate reads into load-based rebalancing (important for ycsb)
* (future work) try to balance reads-per-node when doing leaseholder rebalancing

---

Review status: 0 of 18 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-:-KolyVroB7VkAdWiu6BN:bz8706t)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
After a bunch more testing and playing around with details, I'm not super happy with where this is at.

Fluctuations in writes per second is the main problem. Under steady load, things work fine, but changes in load can have pretty adverse effect on things, and these fluctuations are more common than you might expect -- they're even noticeable on a cluster with no external load. This necessitates a larger difference-from-the-mean threshold before taking action, but that should mostly handle natural fluctuations.

And that handles natural write fluctuations reasonably, but the problem is that there's more to deal with than just natural fluctuations. When a range is split or moved to a new store, we get a replica (or two replicas in the case of a split) with no historical stats about the load on it. If we take a measurement shortly after its stats are initialized, we can get really anomalous QPS numbers due to the short measurement window, both on the high and low end, which cause rebalances that we don't really want to happen. And these measurement spikes are visible not just at the range level, but also at the store level (as in they're very visible in the admin UI graphs), meaning that even if we just don't consider rebalancing a range that was just relocated or split, those anomalous stats can mess with the overall store's QPS that gets gossiped, which can still affect rebalancing decisions about other ranges.

We need some way of smoothing this out. I tried just not counting a replica's QPS at all for 30 seconds, but this led to noticeable drops in QPS on splits and rebalances, which also led to undesired rebalances. Do you have any thoughts on other approaches to take here to avoid affecting store-level stats?

Orthogonally, there's also the fact that the logic as written won't handle extremely skewed workloads well. If one range gets more writes than other entire nodes do, we could end up rebalancing that range onto one of those other nodes in an attempt to even out the QPS. This should be easier to fix, though, by considering what the balance of things will be *after* moving the range, given that we know all the numbers. That's something I should have done in the first place.

---

Review status: 0 of 18 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-:-KpFtjhOGBzJZazQpil3:bic2di5)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
> Do you have any thoughts on other approaches to take here to avoid affecting store-level stats?

For splits, instead of starting the stats from scratch after the spilt, how about copying the stats form the original range for both of them?  That should smooth the stats out when splitting and they should naturally adjust afterwards.  Or even better, split the stats in two for both.

For adding a replica to a new store, can we send the stats with the replica as well.  Even if the historical data are slightly old, having something there is better than starting at 0.  How big are those stats?  Can we attach them to the snapshot perhaps?

---

Review status: 0 of 18 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-:-KpHCPzhCb4Sktd81zX9:bxbox8n)*
<!-- Sent from Reviewable.io -->

petermattis commented:
> Fluctuations in writes per second is the main problem. Under steady load, things work fine, but changes in load can have pretty adverse effect on things, and these fluctuations are more common than you might expect -- they're even noticeable on a cluster with no external load. This necessitates a larger difference-from-the-mean threshold before taking action, but that should mostly handle natural fluctuations.

We might want to use a less sensitive measure of load. There is no need to balance based on load even when there is imbalanced load if doing so is not expected to improve performance because the ranges in question are "keeping" up. Rather than balancing on writes/sec we could use something like queue-length (i.e. the number of pending proposals per range). The thinking here is that a queue length of 0 or 1 would indicate that the node is keeping up with the load while a larger queue length would indicate that rebalancing could be beneficial.

petermattis commented:
It might be worth reading about [queueing theory](https://en.wikipedia.org/wiki/Queueing_theory). Lots of theory and practice about optimizing queueing systems. For example, the [drift plus penalty method](https://en.wikipedia.org/wiki/Drift_plus_penalty) for optimizing time average throughput. I have only read enough about this to be dangerous.

petermattis commented:
PS I know this work is scheduled for 1.1, but I'm willing to re-evaluate that if an obviously good approach does not materialize.

a-robinson committed:
storage: Re-gossip store capacity on significant QPS changes

a-robinson commented:
> For splits, instead of starting the stats from scratch after the spilt, how about copying the stats form the original range for both of them? That should smooth the stats out when splitting and they should naturally adjust afterwards. Or even better, split the stats in two for both.

Yeah, that's what I was testing out yesterday (the splitting in two). It appears to make a pretty significant different in settling things down much better than before.

> For adding a replica to a new store, can we send the stats with the replica as well. Even if the historical data are slightly old, having something there is better than starting at 0. How big are those stats? Can we attach them to the snapshot perhaps?

That's a pretty awkward change that I'm hoping we can do without, but we'll see.

> We might want to use a less sensitive measure of load. There is no need to balance based on load even when there is imbalanced load if doing so is not expected to improve performance because the ranges in question are "keeping" up. Rather than balancing on writes/sec we could use something like queue-length (i.e. the number of pending proposals per range). The thinking here is that a queue length of 0 or 1 would indicate that the node is keeping up with the load while a larger queue length would indicate that rebalancing could be beneficial.

I do like the idea of rebalancing when we're noticing a backup, but is queue length going to be less bursty? And will queues fill up due to system load or just due to too much load on the one range? I don't have a great intuition for what these queues look like in practice.

My thinking had been taking me somewhere slightly different -- specifically, that we should consider more than just the most recent data point when making decisions, but the overall trends and historical data points. In many systems, this would be tough to do, but we already have a timeseries database with historical metrics embedded in the system itself (with the caveat that users can modify the sampling period). If we were to avoid making decisions based on QPS until there had been a steady difference in QPS for a longer time period, we should be able to add some more stability.

a-robinson committed:
storage: Split up write stats when splitting a range

Resetting stats has a bad effect on rebalancing, so avoid doing so when
possible.

a-robinson commented:
I've pushed my latest code after the split change. It performs well with 4 nodes against

1. no load
2. default allocsim load
3. ycsb load

That's obviously a small subset, but it's the best across-the-board performance I've seen so far while trying to keep a fairly tight bound on replica count per store.

I still need to:

1. Add a switch to disable the new dimensions
2. Add logic to check what the balance will be after moving the range (to avoid moving extreme outlier ranges, as explained in https://github.com/cockroachdb/cockroach/pull/16878#issuecomment-315781731). I expect this will involve considerable changes to `shouldRebalance`.
3. Put info on decisions into the range log (or some log somewhere)
4. Fix the existing tests

And that's assuming that we don't switch to trying to use historical data or some other bigger high level change.

a-robinson commented:
Review status: 0 of 21 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/roachpb/metadata.proto, line 96 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-KoNH2GjmDEHvWELazcB-r2-96:-KpLhf7z5g-jLDpRQICo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58a7038f9f572ea91bd355fa8ad81e37ef9967/pkg/roachpb/metadata.proto#L96)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

s/some sort of/the/
</blockquote></details>

Done.

---

*[pkg/storage/allocator.go, line 46 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-KoNH2GkYfZlfyE5u00X-r2-46:-KpLhe5KIy-B0UoC7w3i:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58a7038f9f572ea91bd355fa8ad81e37ef9967/pkg/storage/allocator.go#L46)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

s/baselease/baseLease/
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878)*
<!-- Sent from Reviewable.io -->

petermattis commented:
> I do like the idea of rebalancing when we're noticing a backup, but is queue length going to be less bursty? And will queues fill up due to system load or just due to too much load on the one range? I don't have a great intuition for what these queues look like in practice.

No, I don't think queue length will be less bursty. Certainly we'll need smoothing regardless of the metric. I think we need to rebalance on a metric that indicates overload (or potential overload) rather than one that can vary with no untoward effects. Consider a range that is processing 100 ops/sec while other ranges in the cluster are processing 1 op/sec. Do we need to rebalance so that one node contains the busy range while the other nodes contain the idle ranges? I think it depends on if that range is keeping up. For example, perhaps the 100 ops/sec range is handling 1 request every 10ms and is retiring that request in less than 10ms. 

Btw, I'm not actually sure how I'd measure queue length. Number of pending proposals + number of commands in the command queue. Or perhaps the number of keys in pending proposals and in the command queue. Seems like we could only measure this on the leaseholder.

> If we were to avoid making decisions based on QPS until there had been a steady difference in QPS for a longer time period, we should be able to add some more stability.

Yes, we need to balance between being responsive to load and ignoring short term burstiness.

a-robinson commented:
I'm not yet persisting info about each decision anywhere, but the code here is ready for review.

BramGruneir commented:
Code looks good.  Just some minor comments.

Probably a good idea to run the nightly allocator tests to make sure they still work as expected.

---

Reviewed 2 of 12 files at r1, 10 of 21 files at r3, 11 of 11 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/roachpb/metadata.go, line 254 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-Kpv0ejkMmQ6c2vZKKla:-Kpv0ejkMmQ6c2vZKKlb:b5b9m96) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef819137e5bc0a64a86f4f3983f3560489f316a/pkg/roachpb/metadata.go#L254)):*
> ```Go
> 		return data[len(data)-1]
> 	}
> 	// TODO: Import round function from library?
> ```

Change this todo to use go's rounding functions once it's available?

---

*[pkg/storage/allocator_scorer.go, line 60 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-Kpv30hXayWO-Yv95r9G:-Kpv30hXayWO-Yv95r9H:bma9vd6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef819137e5bc0a64a86f4f3983f3560489f316a/pkg/storage/allocator_scorer.go#L60)):*
> ```Go
> 	//
> 	// TODO(a-robinson): Should disk usage be held to a higher standard than this?
> 	statRebalanceThreshold = 0.20
> ```

Since it seems like these values, `rangeRebalanceThreshold` and `statRebalanceThreshold` might need to be tinkered with, I think making them settings in warranted.

---

*[pkg/storage/allocator_scorer.go, line 518 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-Kpv3s2RREJiXiPwCioM:-Kpv3s2RREJiXiPwCioN:b-v2hnpl) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef819137e5bc0a64a86f4f3983f3560489f316a/pkg/storage/allocator_scorer.go#L518)):*
> ```Go
> 	ctx context.Context, store roachpb.StoreDescriptor, sl StoreList, rangeInfo RangeInfo,
> ) bool {
> 	// TODO(peter,bram,cuong): The FractionUsed check seems suspicious. When a
> ```

I think it's finally time to get rid of this todo!  

---

*[pkg/storage/allocator_scorer.go, line 729 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-Kpv5A68DGaGuReknBaU:-Kpv5A68DGaGuReknBaV:bam9ksz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef819137e5bc0a64a86f4f3983f3560489f316a/pkg/storage/allocator_scorer.go#L729)):*
> ```Go
> 		dimensions.ranges = balanced
> 	}
> 	if EnableStatsBasedRebalancing.Get() {
> ```

Is this check needed?  Aren't we only in here if it is enabled?

---

*[pkg/storage/allocator_scorer.go, line 831 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-Kpv5x_mb9LpJfqoSZsN:-Kpv5x_mb9LpJfqoSZsO:b-h6bwuk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef819137e5bc0a64a86f4f3983f3560489f316a/pkg/storage/allocator_scorer.go#L831)):*
> ```Go
> }
> 
> func rangeIsGoodFit(bd balanceDimensions) bool {
> ```

Probably worth commenting on these, just to be extra clear.

---

*[pkg/storage/allocator_scorer.go, line 890 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-Kpzp5ANieFYgbaQ7eqF:-Kpzp5ANieFYgbaQ7eqG:bkhw3qo) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef819137e5bc0a64a86f4f3983f3560489f316a/pkg/storage/allocator_scorer.go#L890)):*
> ```Go
> }
> 
> func rebalanceToConvergesOnMean(sl StoreList, sc roachpb.StoreCapacity, rangeInfo RangeInfo) bool {
> ```

Any way to combine `rebalanceFromConvergesOnMean` and `rebalanceToConvergesOnMean`?  Looks like it's just three lines that are different.

---

*[pkg/storage/allocator_scorer_test.go, line 706 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-KpzpWbXB64wzoBTGZFI:-KpzpWbXB64wzoBTGZFJ:b-9j7eh3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef819137e5bc0a64a86f4f3983f3560489f316a/pkg/storage/allocator_scorer_test.go#L706)):*
> ```Go
> 
> 	testCases := []struct {
> 		sl       StoreList
> ```

If all test cases use the same storelist, why do you have to specify it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-:-KpzqlqYsKzsMgiv6cgd:bwr94bz)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: 8 of 24 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/roachpb/metadata.go, line 254 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-Kpv0ejkMmQ6c2vZKKla:-KqEknLdc4rXzFvN94nZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef819137e5bc0a64a86f4f3983f3560489f316a/pkg/roachpb/metadata.go#L254)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Change this todo to use go's rounding functions once it's available?
</blockquote></details>

Done.

---

*[pkg/storage/allocator_scorer.go, line 621 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-KoOPp4oq5Mg_rP2bfGk:-KqJnkFWzR19OffRklhr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e58a7038f9f572ea91bd355fa8ad81e37ef9967/pkg/storage/allocator_scorer.go#L621)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm having trouble coming up for intuition for how this behaves. Specifically, `balanceContribution` is a bit ad-hoc with a lot of magic values. I don't have anything better to offer at the moment, though.

Perhaps it would help to have some table driven tests that would demonstrate how `balanceScore` changes as one of the dimensions changes. Even better if we could figure out how to visualize this. 
</blockquote></details>

Done.

---

*[pkg/storage/allocator_scorer.go, line 60 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-Kpv30hXayWO-Yv95r9G:-KqJni7PqWl_zAu-30If:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef819137e5bc0a64a86f4f3983f3560489f316a/pkg/storage/allocator_scorer.go#L60)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Since it seems like these values, `rangeRebalanceThreshold` and `statRebalanceThreshold` might need to be tinkered with, I think making them settings in warranted.
</blockquote></details>

Done.

---

*[pkg/storage/allocator_scorer.go, line 518 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-Kpv3s2RREJiXiPwCioM:-KqEseVehk28pCAjDNTi:bmfxgaw) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef819137e5bc0a64a86f4f3983f3560489f316a/pkg/storage/allocator_scorer.go#L518)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

I think it's finally time to get rid of this todo!  
</blockquote></details>

Done :)

---

*[pkg/storage/allocator_scorer.go, line 729 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-Kpv5A68DGaGuReknBaU:-KqJnsECTC9j_Ql557Vr:bgkcwdt) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef819137e5bc0a64a86f4f3983f3560489f316a/pkg/storage/allocator_scorer.go#L729)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Is this check needed?  Aren't we only in here if it is enabled?
</blockquote></details>

It's needed. This gets called from `allocateCandidates`, `removeCandidates`, and `rebalanceCandidates` without any sort of check guarding the call.

---

*[pkg/storage/allocator_scorer.go, line 831 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-Kpv5x_mb9LpJfqoSZsN:-KqJr8L2I4QPs9UiKnGe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef819137e5bc0a64a86f4f3983f3560489f316a/pkg/storage/allocator_scorer.go#L831)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Probably worth commenting on these, just to be extra clear.
</blockquote></details>

Done.

---

*[pkg/storage/allocator_scorer.go, line 890 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-Kpzp5ANieFYgbaQ7eqF:-KqJtM5jqaU4_DoFhvBx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef819137e5bc0a64a86f4f3983f3560489f316a/pkg/storage/allocator_scorer.go#L890)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Any way to combine `rebalanceFromConvergesOnMean` and `rebalanceToConvergesOnMean`?  Looks like it's just three lines that are different.
</blockquote></details>

Done.

---

*[pkg/storage/allocator_scorer_test.go, line 706 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-KpzpWbXB64wzoBTGZFI:-KqEt1S_fzE1icAd34Tc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef819137e5bc0a64a86f4f3983f3560489f316a/pkg/storage/allocator_scorer_test.go#L706)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

If all test cases use the same storelist, why do you have to specify it?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878)*
<!-- Sent from Reviewable.io -->

a-robinson committed:
roachpb: Improve printing of StoreCapacity

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Just some nits and possible cleanups for a follow-on.

---

Review status: 8 of 24 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/roachpb/metadata.proto, line 103 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-KqQXVbiK6iZc5FEmYJP:-KqQXVbiK6iZc5FEmYJQ:b6cjcki) ([raw file](https://github.com/cockroachdb/cockroach/blob/25757125dc68d304ceb06947bed072f0261cc16c/pkg/roachpb/metadata.proto#L103)):*
> ```protobuf
>   optional double p50 = 3 [(gogoproto.nullable) = false];
>   optional double p75 = 4 [(gogoproto.nullable) = false];
>   optional double p90 = 5 [(gogoproto.nullable) = false];
> ```

It is mildly odd that the size of the percentile "buckets" isn't fixed. Would it make sense to use p16, p33, p50, p66, p83 instead? Probably doesn't matter.

---

*[pkg/roachpb/metadata.proto, line 122 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-KqQZKArzADXJQn0mAQQ:-KqQZKArzADXJQn0mAQR:bx5vhq) ([raw file](https://github.com/cockroachdb/cockroach/blob/25757125dc68d304ceb06947bed072f0261cc16c/pkg/roachpb/metadata.proto#L122)):*
> ```protobuf
>   // This information can be used for rebalancing decisions.
>   optional Percentiles bytes_per_replica = 6 [(gogoproto.nullable) = false];
>   optional Percentiles writes_per_replica = 7 [(gogoproto.nullable) = false];
> ```

Are there any concerns during rolling upgrades to 1.1 about some nodes provides these fields and other nodes not? Should the new rebalancing heuristics hook into the version upgrade machinery?

---

*[pkg/storage/allocator_scorer.go, line 716 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-KqQZ4mTJRxzdp0kUJK9:-KqQZ4mTJRxzdp0kUJKA:b-nyecgh) ([raw file](https://github.com/cockroachdb/cockroach/blob/25757125dc68d304ceb06947bed072f0261cc16c/pkg/storage/allocator_scorer.go#L716)):*
> ```Go
> 
> func oppositeStatus(rcs rangeCountStatus) rangeCountStatus {
> 	return rcs * -1
> ```

`-rcs`?

---

*[pkg/storage/allocator_scorer.go, line 724 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-KqQYwLt_62uEPOk6OUq:-KqQYwLt_62uEPOk6OUr:b-gq3u77) ([raw file](https://github.com/cockroachdb/cockroach/blob/25757125dc68d304ceb06947bed072f0261cc16c/pkg/storage/allocator_scorer.go#L724)):*
> ```Go
> //
> // TODO(a-robinson): Give balanceScore a type alias to discourage direct usage
> // of them except through helper functions?
> ```

Is this TODO still true now that you're returning `balanceDimensions`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-:-KqQZVwi0k0OrEghYWS8:bzhy26r)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Just a question on having this enabled by default.

---

Reviewed 4 of 15 files at r6, 12 of 12 files at r7.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/storage/allocator_scorer.go, line 52 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-KqQql1zjJq_dxhHqExA:-KqQql1zjJq_dxhHqExB:b3cibst) ([raw file](https://github.com/cockroachdb/cockroach/blob/25757125dc68d304ceb06947bed072f0261cc16c/pkg/storage/allocator_scorer.go#L52)):*
> ```Go
> 		"kv.allocator.stats_based_rebalancing.enabled",
> 		"set to enable rebalancing of range replicas based on write load and disk usage",
> 		true)
> ```

Do we want to enable this by default yet? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-:-KqQr1OFzFvQ3gRqiXPZ:b9k7xfp)*
<!-- Sent from Reviewable.io -->

a-robinson committed:
storage: Separate out allocator's convergesScore from constraintScore

I was observing on indigo two replicas of the same range ending up in
the same locality even though there was a third locality without any
copies. This should prioritize diversity more highly.

a-robinson commented:
Review status: 20 of 24 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/roachpb/metadata.proto, line 103 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-KqQXVbiK6iZc5FEmYJP:-KqW1XWk-DokCaQ2eGuv:b-99gij3) ([raw file](https://github.com/cockroachdb/cockroach/blob/25757125dc68d304ceb06947bed072f0261cc16c/pkg/roachpb/metadata.proto#L103)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

It is mildly odd that the size of the percentile "buckets" isn't fixed. Would it make sense to use p16, p33, p50, p66, p83 instead? Probably doesn't matter.
</blockquote></details>

Maybe? Probably?

There also may be something to be said for having smaller buckets toward the more extreme ends of the spectrum, though, such that when we want to move an outlier we're more likely to pick a stronger outlier.

---

*[pkg/roachpb/metadata.proto, line 122 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-KqQZKArzADXJQn0mAQQ:-KqVs6kMRivlxju3QEF6:b-2tn9a2) ([raw file](https://github.com/cockroachdb/cockroach/blob/25757125dc68d304ceb06947bed072f0261cc16c/pkg/roachpb/metadata.proto#L122)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Are there any concerns during rolling upgrades to 1.1 about some nodes provides these fields and other nodes not? Should the new rebalancing heuristics hook into the version upgrade machinery?
</blockquote></details>

Yeah, there are legitimate concerns there. I had been thinking about adding some logic to fall back to the old logic if a cluster had nodes with no percentiles populated, but hooking into the `use_version` from the recent migration discussions would be much simpler.

I'll think about it a little more, but am leaning against changing it right now.

---

*[pkg/storage/allocator_scorer.go, line 52 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-KqQql1zjJq_dxhHqExA:-KqUUEhbq-nB4gGvlImA:bky0b6o) ([raw file](https://github.com/cockroachdb/cockroach/blob/25757125dc68d304ceb06947bed072f0261cc16c/pkg/storage/allocator_scorer.go#L52)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Do we want to enable this by default yet? 
</blockquote></details>

I'd say yes, to bake it as much as possible over the next few weeks so that we can make an informed decisions about whether it needs to be disabled.

---

*[pkg/storage/allocator_scorer.go, line 716 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-KqQZ4mTJRxzdp0kUJK9:-KqUUWCgpdwbLleViGMe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/25757125dc68d304ceb06947bed072f0261cc16c/pkg/storage/allocator_scorer.go#L716)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`-rcs`?
</blockquote></details>

Done.

---

*[pkg/storage/allocator_scorer.go, line 724 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878#-KqQYwLt_62uEPOk6OUq:-KqUUUxbG9dWT5PmZvI8:b-viwntl) ([raw file](https://github.com/cockroachdb/cockroach/blob/25757125dc68d304ceb06947bed072f0261cc16c/pkg/storage/allocator_scorer.go#L724)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is this TODO still true now that you're returning `balanceDimensions`?
</blockquote></details>

Nope, good catch.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 1 of 12 files at r7.
Review status: 20 of 24 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16878)*
<!-- Sent from Reviewable.io -->

knz reviewed (approved):
(checked that the new cluster setting shows up in sql logic tests)

a-robinson commented:
To summarize the latest testing on this:

* It passes the 1-to-3 and 3-to-5 nightly tests
* It failed the 1-to-6 large test, but that's also currently failing on master
* It's been running on indigo with smooth balance of replicas and without thrashing for a while now
* I've been playing around with changing zone configs on indigo. It's a bit on the slow end, but I didn't do a control test to compare it to and it seems to just be bottlenecked by the rate of snapshots in the cluster.

I'm going to merge this and follow up with PRs to integrate it with the new migration/version-gating logic and to add more info about decisions to debug pages.
